pkg/controller/component/storaged_updater.go

```diff
@@ -0,0 +1,320 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package component
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/facebook/fbthrift/thrift/lib/go/thrift"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/rand"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	ng "github.com/vesoft-inc/nebula-go/nebula"
+	"github.com/vesoft-inc/nebula-operator/apis/apps/v1alpha1"
+	"github.com/vesoft-inc/nebula-operator/pkg/kube"
+	"github.com/vesoft-inc/nebula-operator/pkg/nebula"
+	utilerrors "github.com/vesoft-inc/nebula-operator/pkg/util/errors"
+	extenderutil "github.com/vesoft-inc/nebula-operator/pkg/util/extender"
+	"github.com/vesoft-inc/nebula-operator/pkg/util/resource"
+)
+
+const (
+	// TransLeaderBeginTime is the key of trans Leader begin time
+	TransLeaderBeginTime = "transLeaderBeginTime"
+	// TransLeaderTimeout is the timeout limit of trans leader
+	TransLeaderTimeout = 3 * time.Minute
+)
+
+type storagedUpdate struct {
+	client.Client
+	clientSet kube.ClientSet
+	extender  extenderutil.UnstructuredExtender
+}
+
+func NewStoragedUpdater(cli client.Client, clientSet kube.ClientSet) UpdateManager {
+	return &storagedUpdate{
+		Client:    cli,
+		clientSet: clientSet,
+		extender:  extenderutil.New(),
+	}
+}
+
+func (s *storagedUpdate) Update(
+	nc *v1alpha1.NebulaCluster,
+	oldUnstruct, newUnstruct *unstructured.Unstructured,
+	gvk schema.GroupVersionKind,
+) error {
+	log := getLog().WithValues("namespace", nc.GetNamespace(), "name", nc.GetName())
+	if *nc.Spec.Storaged.Replicas == int32(0) {
+		return nil
+	}
+
+	if nc.Status.Storaged.Phase == v1alpha1.ScaleInPhase || nc.Status.Storaged.Phase == v1alpha1.ScaleOutPhase {
+		log.Info("storaged is scaling, can not update")
+		return getLastConfig(s.extender, oldUnstruct, newUnstruct)
+	}
+
+	nc.Status.Storaged.Phase = v1alpha1.UpdatePhase
+	if !extenderutil.PodTemplateEqual(s.extender, newUnstruct, oldUnstruct) {
+		return nil
+	}
+
+	if nc.Status.Storaged.Workload.UpdateRevision == nc.Status.Storaged.Workload.CurrentRevision {
+		return nil
+	}
+
+	spec := s.extender.GetSpec(oldUnstruct)
+	actualStrategy := spec["updateStrategy"].(map[string]interface{})
+	advanced := gvk.Kind == resource.AdvancedStatefulSetKind.Kind
+	partition := actualStrategy["rollingUpdate"].(map[string]interface{})
+	if err := setPartition(s.extender, newUnstruct, partition["partition"].(int64), advanced); err != nil {
```

Comment:
Why set to the `oldUnstruct`'s `partition`?

Reply:
It used for controlling update steps.